### ISSUE TITLE
Should we simplify the YAML structure of metrics?

### ISSUE THREAD

**Issue by apannetrat** [cloudsecurityalliance/continuous-audit-metrics#19]
I was trying to document the YAML data structure, and I came across these questions: 

#### 1. Does `samplingPeriod` need to appear both in the parameters of the metric formula and in the metric? 

Shouldn't a single `samplingPeriod` at the metric level be enough?

#### 2. Do the `count` and `type` metric parameter elements provide useful information for the metric?

If yes, we need to explain how these things are defined and used.

#### 3. Isn't `crontabExample` redundant with the `samplingPeriod`.

In many cases, the `crontabExample` is also incorrect, as previously noted.

**Comment by pritikin** [comment]
Addressing the simpler points first: 
3. crontabExample was the prior work where we were using chrontab formats for periods. Rather than remove it immediately we used this intermediate step. I think its a reasonable position that we drop this entirely until such time as CAML ( https://github.com/ContiNube/CAML ) can leverage it. 

2. the unit (e.g. count) and type (calculate) were brought in from the openmetrics/prometheus/grafana space. In our original spreadsheet we had 'category' & 'unit' which are roughly analogous. The common theme is to capture some meta information about a metric that could be determined from reading the more detailed implementation guidelines but could also be more clearly specified. Interestingly in some spaces the idea is to send this type information along with the metric during reporting (presumably so that downstream systems have some concept of how to present this to end users w/o needing to look it up in the catalog or be locally configured). We have added and dropped these types of fields a number of times. 

Both of these are examples of engineering for the future. In principle we could drop them until such time that they're really needed. In the case of unit and type it is to support prometheus (time series database) and grafana (visualization) use cases.

**Comment by pritikin** [comment]
Sampling period discussions might be big enough for its own issue. I continue to find myself confused by metric period discussions and am looking forward to this discussion to solidify my thinking. (Thats a request for folks to *correct* me where i've gone astray! I feel like I'm just thinking outloud here)

The 'parameters' are what raj has referred to as measures. These are specific measurements like "Total Number of deployed production applications" and are inputs into the numerator/denominator of the metric expression. 

For some measures/parameters the meaning is clear. If we look at BCR-06-M1 ("reports the percentage of critical systems that passed BCR tests") each measure is a count: 

>   metricDescription: This metric reports the percentage of critical systems that passed
>     BCR tests.
>   expression:
>     formula: "(A/B)*100"
>     parameters:
>     - id: bcr_tests_critical_systems
>       name: A
>       description: Number of critical systems that passed BCR tests during the sampling
>         period
>       unit: count
>       type: calculate
>       samplingPeriod: P365D
>     - id: critical_systems_count
>       name: B
>       description: Total number of critical systems operating during the sampling
>         period
>       unit: count
>       type: calculate
>       samplingPeriod: P365D

If the implementation model is "some script collects the measures and computes the metric" then it makes sense to have a single samplePeriod set at the metric level. The measures/parameters would inherit from the metric level. The script would collect all the data w/in the time period to compute the measure. In our example here it determines the count of critical systems (both 'critical_systems_count' and number of these that passed bcr tests; within the last year).

If the implementation is "time series measurements are taken and stored in the time series database regularly" then the computation of the metric involves finding to correct measure in that database and using it to compute the metric. Now the whole concept of a time period for the metric is either "case1: extra important" or "case2: not important at all". 

Case1, if the measures are collected and stored in a db like described in https://github.com/ContiNube/CAML/tree/dev/docs/img then we need the period along with the measures to be able to compute the metric over the same period. I think Raj could provide input on how continube components might work like this. Going this route requires us to build out an architecture to support our metric mode.

Case2: we really lean into the time series aspect. In this case the time series measures are specific counts (measures) at specific times. Thus the value of labeling them as a type w/ 'count', 'gauge', 'histogram', 'summary' etc. In this case we would query the time series database AND THEN perform the metric calculation. In effect pushing off the 'expression' until needed rather than pre-calculating it and pushing the results into the time series stream.

That case2 discussion might aligns nicely with your original question. Assuming that we can make the measures be counts or sums then we can drop period information. We then compute the metric by querying the time series database for the measures taken during the period. 

In short, I think until we understand which of these models we prefer we won't be able to answer your very important question.

**Comment by pritikin** [comment]
My temptation is to limit the period to the metric in the yaml. As such we should evaluate if there is any case where we indicate different values. Some quick notes: 

CCC-07-M1 is an example where we indicate different periods. 
total_config_items has samplePeriod of P7D but the sloPeriod is P30D.

**Comment by pritikin** [comment]
measurementFrequency is probably part of this conversation as well. For the most part its consistent with the samplingPeriod.

It is missing for TVM-07-M1

**Comment by pritikin** [comment]
We discussed this at the 4/29 WG meeting and achieved general consensus that:
1) for the catalog describing the period only at the metric level is sufficient.
2) the question of period is very important to the implementation of measures (parameters) and needs to be discussed/worked-though in order to see ultimate success. The simplest statement is that parameters must 'inherit' from the metric; but this does not imply a decision in the time series vs non-time series database choices.

**Comment by pritikin** [comment]
5/6 - one more PI needed to strip the count & type fields

**Comment by pritikin** [comment]
closing this as it appears we resolved the immediate concerns.
